### PR TITLE
Add syntax support for CREATE MATERIALIZED VIEW

### PR DESCRIPTION
**Pull request by gggrace14**
Make Presto parser support CREATE MATERIALIZED VIEW syntax, ex.

CREATE MATERIALIZED VIEW IF NOT EXISTS mv
COMMENT 'A simple materialized view'
WITH (
	partitioned_by = ARRAY['ds'],
	retention_days = 90
)
AS SELECT * FROM t

We do not support REPLACE MATERIALIZED VIEW at the moment,
since that is possible to trigger the entire view refresh.
Users could use DROP and CREATE a new materialzied view instead.

This commit rewrites a subset of
prestosql/presto@88116a4
due to above reason.

Test plan -
1. Add unit tests to TestSqlParser
2. mvn test

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by kaikalur** [comment]
High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.

**Comment by jainxrohit** [comment]
> High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.

@kaikalur I think its good to have a separate statement. This creates the visibility to the users. As its very different from the logical view. A similar approach was taken by the prestosql
https://github.com/prestosql/presto/pull/3283/files

**Comment by jainxrohit** [comment]
> This commit rewrites a subset of

I would say lets come with a document explaining more examples as in how it would be used?

**Comment by kaikalur** [comment]
> > High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.
> 
> @kaikalur I think its good to have a separate statement. This creates the visibility to the users. As its very different from the logical view. A similar approach was taken by the prestosql
> https://github.com/prestosql/presto/pull/3283/files

I'm not a fan of that approach - apart from the fact that it's materialized all the other aspects remain the same as a view so why add all the visitors and code bloat and duplication everywhere.

**Comment by gggrace14** [comment]
> > > High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.
> > 
> > 
> > @kaikalur I think its good to have a separate statement. This creates the visibility to the users. As its very different from the logical view. A similar approach was taken by the prestosql
> > https://github.com/prestosql/presto/pull/3283/files
> 
> I'm not a fan of that approach - apart from the fact that it's materialized all the other aspects remain the same as a view so why add all the visitors and code bloat and duplication everywhere.

Hi @kaikalur , one key motivation is that we need to bind the CreateMaterializedView Statement to CreateMaterializedViewTask which would be a quite different implementation from CreateViewTask. No way to merge logic into CreateViewTask.

Also we could barely support the complete syntax of CreateView (copied below) for CreateMaterializedView, ex. "REPLACE", "SECURITY (DEFINER | INVOKER)". Some are not practical for MV. So making MATERIALIZED an attribute of CreateView statement might not work in our case.

We don't want to mess up the CreateView logic. It's more clean and modularized to keep CreateMaterializedView separate.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java
Comment (highker) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java:39]: remove this as we don't use it

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java
Comment (highker) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java:37]: I would move this to a static helper in the test

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java
Comment (highker) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CreateMaterializedView.java:51]: requireNonNull

**Comment by jainxrohit** [comment]
@gggrace14 I tried a simple query, however it fails, am I missing something?

presto:tpch> CREATE MATERIALIZED VIEW test_mv_nation AS select * from nation;
Query 20201021_043817_00000_bramu failed: Unsupported statement type: CreateMaterializedView
com.facebook.presto.spi.PrestoException: Unsupported statement type: CreateMaterializedView
        at com.facebook.presto.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:113)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

**Review by jainxrohit — CHANGES_REQUESTED** [review]
It seems things are not working with current changes?

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java
Comment (highker) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1497]: Use `ImmutableList.of()` instead of `new ArrayList<>()`; same for other places

**Review by highker — COMMENTED** [review]
LGTM % nits

**Comment by kaikalur** [comment]
> > > > High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.
> > > 
> > > 
> > > @kaikalur I think its good to have a separate statement. This creates the visibility to the users. As its very different from the logical view. A similar approach was taken by the prestosql
> > > https://github.com/prestosql/presto/pull/3283/files
> > 
> > 
> > I'm not a fan of that approach - apart from the fact that it's materialized all the other aspects remain the same as a view so why add all the visitors and code bloat and duplication everywhere.
> 
> Hi @kaikalur , one key motivation is that we need to bind the CreateMaterializedView Statement to CreateMaterializedViewTask which would be a quite different implementation from CreateViewTask. No way to merge logic into CreateViewTask.
> 
> Also we could barely support the complete syntax of CreateView (copied below) for CreateMaterializedView, ex. "REPLACE", "SECURITY (DEFINER | INVOKER)". Some are not practical for MV. So making MATERIALIZED an attribute of CreateView statement might not work in our case.
> 
> CREATE (OR REPLACE)? VIEW qualifiedName
> (SECURITY (DEFINER | INVOKER))? AS query
> 
> We don't want to mess up the CreateView logic. It's more clean and modularized to keep CreateMaterializedView separate.

Then how is it different from CTAS? Also, implementation being different is fine. I was talking about analysis - the analysis you do for views should be the same - like access rules etc?

**Comment by rongrong** [comment]
I kinda agree with @kaikalur here. Does it make sense to make materialized view a property? And independent of this, I actually prefer to handle `CREATE VIEW` and `CREATE MATERIALIZED VIEW` together. As Sreeni mentioned, analysis / security checks on both should be the same. The two should share a lot of code at AST analyzer level. So even if we add the `MATERIALIZED` keyword, I still prefer to update it to `CREATE (OR REPLACE)? (MATERIALIZED)? VIEW`.

**Comment by gggrace14** [comment]
> @gggrace14 I tried a simple query, however it fails, am I missing something?
> 
> presto:tpch> CREATE MATERIALIZED VIEW test_mv_nation AS select * from nation;
> Query 20201021_043817_00000_bramu failed: Unsupported statement type: CreateMaterializedView
> com.facebook.presto.spi.PrestoException: Unsupported statement type: CreateMaterializedView
> at com.facebook.presto.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:113)
> at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
> at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
> at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

Hi @jainxrohit this is intended behavior originally. It throws because I didn't bind the CreateMaterializedView statement to any Task, as in https://github.com/prestodb/presto/blob/051f7b658c6b28ba4cbb248a7ad3d9d9f2f01d9c/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L330. With this change, "Create Materialized View" is supposed not to do anything. A similar change is https://github.com/prestodb/presto/commit/0ec662197265fe75eeef2e54d74c0a2eb552a568. If you prefer not to print message explicitly, I could bind the CreateMaterializedView statement to a Task that does nothing. Let me know.

**Comment by gggrace14** [comment]
> > > > > High level comment - can we just make MATERIALIZED an attribute of the ViewStatement? For all other purposes, it's just a like a view. So i don't think a separate Statement class is needed. Just add the attribute to View and just use that in the plan generation.
> > > > 
> > > > 
> > > > @kaikalur I think its good to have a separate statement. This creates the visibility to the users. As its very different from the logical view. A similar approach was taken by the prestosql
> > > > https://github.com/prestosql/presto/pull/3283/files
> > > 
> > > 
> > > I'm not a fan of that approach - apart from the fact that it's materialized all the other aspects remain the same as a view so why add all the visitors and code bloat and duplication everywhere.
> > 
> > 
> > Hi @kaikalur , one key motivation is that we need to bind the CreateMaterializedView Statement to CreateMaterializedViewTask which would be a quite different implementation from CreateViewTask. No way to merge logic into CreateViewTask.
> > Also we could barely support the complete syntax of CreateView (copied below) for CreateMaterializedView, ex. "REPLACE", "SECURITY (DEFINER | INVOKER)". Some are not practical for MV. So making MATERIALIZED an attribute of CreateView statement might not work in our case.
> > CREATE (OR REPLACE)? VIEW qualifiedName
> > (SECURITY (DEFINER | INVOKER))? AS query
> > We don't want to mess up the CreateView logic. It's more clean and modularized to keep CreateMaterializedView separate.
> 
> Then how is it different from CTAS? Are we going to automatically update the view when any of the tables changes? I just want to make sure we are not adding something
> 
> Also, implementation being different is fine. I was talking about analysis - the analysis you do for views should be the same - like access rules etc?

Hi @kaikalur one way I see it different from CTAS is that CTAS involves a query execution and populates the table right away. However, CreateMaterializedView just creates the table and doesn't populate it. We would have a separate workflow (running Chronos job actually) to populate it on demand. So CTAS is bound to SqlQueryExecution, while CreateMaterializedView would be bound to DataDefinitionExecution.

Also CMV will need to write quite a bit more metadata to HMS. Ex., links from the base table to MV, needed for query rewrite. We have design doc here, if it's okay to link it to GitHub https://fb.quip.com/ZdrfAF313Mco.

In terms of access rules, it is to be decided at the moment. It's likely to go more complex than CV, as discussed here https://docs.google.com/document/d/1jHS4Bgh63jTmYo_osDZJu7JX8jUQmcFhTTlXscVR0EQ/edit?disco=AAAAIwLo_AI.

My understanding is, if we will have different implementations, there's no way to bind the same Statement to different Tasks in here, https://github.com/prestodb/presto/blob/cf2213e9960f72f90ac87fd182d45bc30a2e9f74/presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java#L330.

**Review by highker — APPROVED** [review]

**Comment by highker** [comment]
@kaikalur, @rongrong, materialized view semantics and view semantics are very different. A view is nothing but a syntactic sugar by helping users to simplify the SQL authoring. Materialized views are more like a table. Especially, they are CTAS tables. The only difference between CTAS vs materialized views are that a CTAS is a table but a materialized view is "view" on the outside but table as the inside. The ACL should respect the underlying base tabes, which is different from today's view ACL. Also, materialized views also need to get the views updated if the underlying tables are updated. For example, new partition lands, remove partition if it's out of retention, etc.

**Review by jainxrohit — APPROVED** [review]

**Comment by gggrace14** [comment]
> @kaikalur, @rongrong, materialized view semantics and view semantics are very different. A view is nothing but a syntactic sugar by helping users to simplify the SQL authoring. Materialized views are more like a table. Especially, they are CTAS tables. The only difference between CTAS vs materialized views are that a CTAS is a table but a materialized view is "view" on the outside but table as the inside. The ACL should respect the underlying base tabes, which is different from today's view ACL. Also, materialized views also need to get the views updated if the underlying tables are updated. For example, new partition lands, remove partition if it's out of retention, etc.

Thank you @highker for inputs to the question!
Thank @highker @jainxrohit @kaikalur @rongrong for reviewing my PR!

**Comment by highker** [comment]
I will merge the PR but wanna get @rongrong and @kaikalur's opinion on it after we explain the relationale behind materialized views.

**Comment by kaikalur** [comment]
> Also, materialized views also need to get the views updated if the underlying tables are updated. For example, new partition lands, remove partition if it's out of retention, etc.

This is the part that makes it different. Otherwise, we might as well just do CTAS. I'm ok now.

**Review by kaikalur — APPROVED** [review]

**Review by rongrong — APPROVED** [review]

**Comment by kaikalur** [comment]
I think this requires more discussion with respect to syntactic/semantic issues. So I say we should hold off merging for now.

**Review by kaikalur — CHANGES_REQUESTED** [review]
See my last comment. Let's discuss it more before merging.

**Comment by gggrace14** [comment]
> See my last comment. Let's discuss it more before merging.
Hi @kaikalur 
I scheduled a short meeting tomorrow for us to discuss it and the context.

**Comment by highker** [comment]
Chatted offline with @rongrong, we will build materialized views with the following two requirements:
- Consistency: Materialized view is a view. When a materialized view is created, no matter whether the materialized data is up to date or not, querying the materialized view should return the exact result as if it's a non-materialized view. This can be done through stitching, meaning that if the some partition data is up to date, the result will leverage the materialized part; for the remaining part, the result should stitch from the base table based on the view query.
- Query optimization: If a query or part of a query against the source table can instead be resolved by querying the materialized view, the optimizer will rewrite the query to use the materialized view for better performance and efficiency.

If there is no objection, I will merge this PR and continue the path with the above requirements. There will be major changes to our read path and minor changes to connector API on write path.

**Pull request merged**
